File: basic_dsa_two_hundred_eighty_two.java
/**
 * Basic DSA Problems - Invert Binary Tree
 * Swap left and right subtrees recursively
 */
public class basic_dsa_two_hundred_eighty_two {

    // Problem: Invert the binary tree (swap left and right of every node).
    // Return root.

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int v) { val = v; }
    }

    /**
     * Recursive: swap left/right, then invert both subtrees.
     * Time O(n), Space O(h)
     */
    public static TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode tmp = root.left;
        root.left = invertTree(root.right);
        root.right = invertTree(tmp);
        return root;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Invert Binary Tree");
        System.out.println("=".repeat(60));

        TreeNode r = new TreeNode(4);
        r.left = new TreeNode(2);
        r.right = new TreeNode(7);
        r.left.left = new TreeNode(1);
        r.left.right = new TreeNode(3);
        r.right.left = new TreeNode(6);
        r.right.right = new TreeNode(9);

        invertTree(r);
        System.out.println("4(2(1,3),7(6,9)) inverted -> 4(7(9,6),2(3,1))");
    }
}

File: basic_dsa_two_hundred_eighty_two.cpp
/**
 * Basic DSA - Balanced Binary Tree (LeetCode 110)
 * Return true if height difference of subtrees <= 1 at every node.
 */
#include <iostream>
#include <algorithm>
#include <cmath>

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int height(TreeNode* root) {
    if (!root) return 0;
    int L = height(root->left);
    int R = height(root->right);
    if (L == -1 || R == -1 || std::abs(L - R) > 1) return -1;
    return 1 + std::max(L, R);
}

bool isBalanced(TreeNode* root) {
    return height(root) != -1;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Balanced Binary Tree\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    std::cout << "[3,9,20,null,15,7] -> " << (isBalanced(root) ? "true" : "false") << "\n";

    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    root2->right = new TreeNode(2);
    root2->left->left = new TreeNode(3);
    root2->left->right = new TreeNode(3);
    root2->left->left->left = new TreeNode(4);
    root2->left->left->right = new TreeNode(4);
    std::cout << "[1,2,2,3,3,null,null,4,4] -> " << (isBalanced(root2) ? "true" : "false") << "\n";
}
